app/src/main/java/com/kamron/pogoiv/ScreenGrabber.java

```diff
@@ -144,4 +114,55 @@ private Bitmap getBitmap(ByteBuffer buffer, int pixelStride, int rowPadding) {
 
         return bmp;
     }
+
+    public @Nullable @ColorInt int[] grabPixels(Point[] points) {
+        Image image = null;
+        try {
+            //Note: imageReader shouldn't be null, but apparently sometimes is.
+            //Let's allow this to still happen.
+            image = imageReader.acquireLatestImage();
+        } catch (Exception exception) {
+            Timber.e("Error thrown in grabPixels() - acquireLatestImage()");
+            Timber.e(exception);
+        }
+
+        if (image == null) {
+            return null;
+        }
+
+        Rect imageBounds = new Rect(0, 0, image.getWidth(), image.getHeight());
+        @ColorInt int[] pixels = new int[points.length];
+
+        Image.Plane imgPlane = image.getPlanes()[0];
+        int pixelStride = imgPlane.getPixelStride();
+        int rowStride = imgPlane.getRowStride();
+        ByteBuffer buffer = imgPlane.getBuffer();
+
+        for (int i = 0; i < points.length; i++) {
+            Point p = points[i];
+            if (imageBounds.contains(p.x, p.y)) {
+                pixels[i] = getPixel(buffer, p, pixelStride, rowStride);
```

Comment:
I'm beginning to wonder if there's any guarantee that this buffer we get from ImageReader is guaranteed to be an uncompressed image. What is the projections writes the image into the ImageReader in PNG format? Then, directly accessing the buffer isn't going to work?

Also, taking a closer look at the Bitmap docs and grabScreen() I think a simpler code with the same benefits (the main benefit of this change is to avoid copyPixelsFromBuffer) would be to use [this constructor](https://developer.android.com/reference/android/graphics/Bitmap.html#createBitmap(int[], int, int, android.graphics.Bitmap.Config)) when creating the bitmap in grabScreen? Since this constructor is supposed to produce and "immutable" Bitmap, I think it doesn't do any buffer copying. The current code needs copyFromBuffer because we first create a mutable Bitmap and then set the pixel values. Thoughts?

While also led me to another confusion. The [Bitmap.createBitmap() calls we use in OcrHelper](https://developer.android.com/reference/android/graphics/Bitmap.html#createBitmap(android.graphics.Bitmap, int, int, int, int)) to cut out the pieces of the bitmap are supposed to return an immutable Bitmap. However, inside replaceColors, we still getPixels, change the values and setPixels? So, how the heck are we able to mutate an immutable Bitmap? How is .setPixels even allowed on an immutable object? Or am I misunderstanding the meaning of immutable?